cloudsync/tasks.py

```diff
@@ -23,58 +27,100 @@
 THUMBNAIL_PATTERN = "thumbnails/{}_thumbnail_{{count}}"
 
 
-@shared_task(bind=True)
-def stream_to_s3(self, url, s3_key):
+class VideoTask(Task):
+    """
+    Custom Celery Task class for video uploads and transcodes
+    """
+
+    def get_task_id(self):
+        """
+        Get the task id (depending on whether the task is chained or not)
+
+        Args:
+            request(Task.request): The task request
+
+        Returns:
+            The request id
+        """
+        if self.request.chain:
+            try:
+                return self.request.chain[0]['options']['task_id']
+            except (IndexError, KeyError,) as exc:
+                # Log the error and continue, using self.request.id instead
+                # The worst that will happen is that progress bar won't work.
+                log.error("Could not find task_id in chain: %s", str(exc))
+        return self.request.id
+
+
+@shared_task(bind=True, base=VideoTask)
+def stream_to_s3(self, video_id):
     """
     Stream the contents of the given URL to Amazon S3
     """
-    if not url:
+
+    if not video_id:
         return False
-    response = requests.get(url, stream=True)
-    response.raise_for_status()
+
+    video = Video.objects.get(id=video_id)
+    video.update_status(VideoStatus.UPLOADING)
+
+    task_id = self.get_task_id()
+    try:
+        response = requests.get(video.source_url, stream=True, timeout=60)
+        response.raise_for_status()
+    except Exception:
+        video.update_status(VideoStatus.UPLOAD_FAILED)
+        self.update_state(task_id=task_id, state=states.FAILURE)
+        raise
+
     _, content_type, content_length = parse_content_metadata(response)
 
     s3 = boto3.resource('s3')
     bucket_name = settings.VIDEO_S3_BUCKET
     bucket = s3.Bucket(bucket_name)
-
-    # Need to bind this here, because otherwise it gets lost in the callback somehow
-    task_id = self.request.id
+    total_bytes_uploaded = 0
 
     def callback(bytes_uploaded):
         """
         Callback function after upload
         """
+        nonlocal total_bytes_uploaded
+        total_bytes_uploaded = total_bytes_uploaded + bytes_uploaded
         data = {
-            "uploaded": bytes_uploaded,
+            "uploaded": total_bytes_uploaded + bytes_uploaded,
```

Comment:
To get the number of bytes transferred since the upload began, so it can be compared to the file size (total), and percentage progress calculated.